python/lsst/summit/utils/consdbClient.py

```diff
@@ -0,0 +1,412 @@
+        """Retrieve the valid keys for a flexible metadata table.
+
+        Parameters
+        ----------
+        instrument: `str`
+            Name of the instrument (e.g. ``LATISS``).
+        obs_type: `str`
+            Name of the observation type (e.g. ``Exposure``).
+
+        Returns
+        -------
+        key_info: `list` [ `list` [ `str` | `None` ] ]
+            List of string lists, one for each key. Each string list contains
+            the name of the key, a data type string, and a documentation
+            string.  A unit string and UCD string may also be present.
+        """
+        url = _urljoin(self.url, "flex", quote(instrument), quote(obs_type), "schema")
+        return self._handle_get(url)
+
+    def get_flexible_metadata(
+        self, instrument: str, obs_type: str, obs_id: int, keys: list[str] | None = None
+    ) -> dict[str, Any]:
+        """Get the flexible metadata for an observation.
+
+        Parameters
+        ----------
+        instrument: `str`
+            Name of the instrument (e.g. ``LATISS``).
+        obs_type: `str`
+            Name of the observation type (e.g. ``Exposure``).
+        obs_id: `int`
+            Unique observation id.
+        keys: `list` [ `str` ], optional
+            List of keys to be retrieved; all if not specified.
+
+        Returns
+        -------
+        result_dict: `dict` [ `str`, `Any` ]
+            Dictionary of key/value pairs for the observation.
+        """
+        url = _urljoin(
+            self.url,
+            "flex",
+            quote(instrument),
+            quote(obs_type),
+            "obs",
+            quote(str(obs_id)),
+        )
+        return self._handle_get(url, {"k": keys} if keys else None)
+
+    def insert_flexible_metadata(
+        self,
+        instrument: str,
+        obs_type: str,
+        obs_id: int,
+        values: dict[str, Any] | None = None,
+        *,
+        allow_update: bool = False,
+        **kwargs,
+    ) -> requests.Response:
+        """Set flexible metadata values for an observation.
+
+        Parameters
+        ----------
+        instrument: `str`
+            Name of the instrument (e.g. ``LATISS``).
+        obs_type: `str`
+            Name of the observation type (e.g. ``Exposure``).
+        obs_id: `int`
+            Unique observation id.
+        values: `dict` [ `str`, `Any` ], optional
+            Dictionary of key/value pairs to add for the observation.
+        allow_update: `bool`, optional
+            If ``True``, allow replacement of values of existing keys.
+        **kwargs: `dict`
+            Additional key/value pairs, overriding ``values``.
+
+        Returns
+        -------
+        response: `requests.Response`
+            HTTP response from the server, with 200 status for success.
+        """
+        if values:
+            values.update(kwargs)
+        else:
+            values = kwargs
```

Comment:
I see how this works, but I'm not sure why you do it like this - I'm more just curious how you see this being used than anything else. Why allow an empty values dict and support kwargs?

Other than the explanation, the only comment would therefore be to add an `if not values: return` or something like that, as this opens up the possibility that there are no kwargs and also no values supplied (maybe with a warning to boot, too, actually).

Reply:
I thought this was convenient and similar to Butler data ids, where you can specify a dict or kwargs, with the latter overriding the former.